pkg/apis/build/v1alpha1/build_validation.go

```diff
@@ -92,3 +97,53 @@ func (bt *BuildSpec) validateTimeout() *apis.FieldError {
 	}
 	return nil
 }
+
+// Validate source
+func (bs BuildSpec) validateSources() *apis.FieldError {
+	var subPathExists bool
+	names := map[string]string{}
+	sourcePath := map[string]string{}
+
+	// both source and sources cannot be defined in build
+	if len(bs.Sources) > 0 && bs.Source != nil {
+		return apis.ErrMultipleOneOf("b.spec.source", "b.spec.sources")
+	}
+
+	for _, source := range bs.Sources {
+		// check all source have unique names
+		if _, ok := names[source.Name]; ok {
+			return apis.ErrMultipleOneOf("b.spec.sources.names")
+		}
+		// multiple sources cannot have subpath defined
+		if source.SubPath != "" {
+			if subPathExists {
+				return apis.ErrInvalidValue("b.spec.sources.subpath", source.SubPath)
+			}
+			subPathExists = true
+		}
+		names[source.Name] = ""
+
+		if err := validatePath(source.TargetPath, sourcePath); err != nil {
+			return err
+		}
+		sourcePath[source.TargetPath] = ""
+
+	}
+	return nil
+}
+
+// validatePath checks that source path is unique
+// or subpath is unique
+func validatePath(path string, pathMap map[string]string) *apis.FieldError {
+	err := apis.ErrMultipleOneOf("b.spec.sources.targetPath")
+
+	if _, ok := pathMap[path]; ok {
+		return err
+	}
+	for key := range pathMap {
```

Comment:
Seems like we can put a `if path == "" { return nil }` before this? 

What behavior do you want when TargetPath == ""?